sklearn/cluster/_birch.py

```diff
@@ -579,10 +581,11 @@ def predict(self, X):
         """
         X = check_array(X, accept_sparse='csr')
         self._check_fit(X)
-        reduced_distance = safe_sparse_dot(X, self.subcluster_centers_.T)
-        reduced_distance *= -2
-        reduced_distance += self._subcluster_norms
-        return self.subcluster_labels_[np.argmin(reduced_distance, axis=1)]
+        with sklearn.config_context(working_memory=16):
```

Comment:
Why such a low working memory ?
Is there something wrong with the default 1GB of working memory ? I think 1GB is reasonable.

Reply:
I think this is the key issue. The Python gc doesent't always release the memory of the distances matrix immediately after a chunk was processed and before the next chunk begins. So we can accumulate several chunk_sizes in memory.
From the performance output we can see that the delta between inital memory and peak reached 70 MB although chunk size is 16MB.

On my macbook i have 8GB of memory. With a chunk size of 1GB if the process accumulates 4 chunks before gc releases, this is 4 GB, and it can have an impact.

Reply:
> With a chunk size of 1GB if the process accumulates 4 chunks before gc releases

Have you experienced that ? I just ran a test with a 8GB pairwise distance matrix several times and I consistently get only 1GB increase of memory. It's possible that with very small chunks the gc is not called at a fast enough rate to collect after each chunk but I'd expect it would be the case for larger chunks.

Reply:
OK. So I"ll stik to the 1GB working_memory then.

Reply:
I'd say so, unless you actually observed that the gc rate is too low even for this chunk size.